### PR TITLE
LibWeb: Layout text chunks based on their Unicode direction

### PR DESCRIPTION
**Pull request by BenJilks**
This pull request is further progress on issue https://github.com/LadybirdBrowser/ladybird/issues/862, all text within a fragment is now rendered RTL.

### CODE REVIEW

**Comment by diegoiast** [comment]
Close - but things are a little more tricky:

image

1. Paragraph direction is not set. Yes, you are aligning to the right on RTL blocks, but the direction of the paragraph shold be set. How can you tell? Look at the "." at the welcome banner.  The "dots" should be at the right part of the sentences - where the end of the RTL paragraph is.
1. Numbers are reversed. Look at the first line or the blue block ("did you know?")  you will see "08" - but in reality it should display "80". Numbers are always  written from left-to-right. Wierdly - I do not see this in Arabic scripts (I see 2283, 1446 properly displayed):
image
1. Selection in CTL scripts (tested hebrew and arabic, Hindi CJK should also be tested eventually) is non existant. Is it a regression, or just not supported on Ladybird's main?


I attach here a small HTML snippet which does the issues with paragraph direction

Firefox (correct)
image

Ladybird (alsmost):
image


```html
<div dir="ltr" align="left">
hello 1, 2, 3!
</div>

<div dir="rtl" align="left">
hello 1, 2, 3!
</div>

<div dir="ltr" align="right">
hello 1, 2, 3!
</div>

<div dir="rtl" align="right">
hello 1, 2, 3!
</div>
```

Ping me in discord, or directly via email if you need more guidance. This is not trivial, and many corners cases need to be fixed.

**Comment by BenJilks** [comment]
Thanks for the feedback. I can't read any RTL scripts, so it's really appreciated. As you mentioned at the end, I’m ignoring layout for now, as that’s a whole other set of issues.

File: Userland/Libraries/LibWeb/Layout/InlineFormattingContext.cpp [hunk 1]
```diff
@@ -389,7 +389,8 @@ void InlineFormattingContext::generate_line_boxes(LayoutMode layout_mode)
                 item.margin_end,
                 item.width,
                 text_node.computed_values().line_height(),
-                move(item.glyph_run));
+                move(item.glyph_run),
+                item.direction);
```
Comment (diegoiast) [inline Userland/Libraries/LibWeb/Layout/InlineFormattingContext.cpp:393]: You should inherit direction from parent . Are you sure this is correct?

File: Userland/Libraries/LibWeb/Layout/LineBox.h [hunk 2]
```diff
@@ -20,7 +20,7 @@ class LineBox {
     CSSPixels bottom() const { return m_bottom; }
     CSSPixels baseline() const { return m_baseline; }
 
-    void add_fragment(Node const& layout_node, int start, int length, CSSPixels leading_size, CSSPixels trailing_size, CSSPixels leading_margin, CSSPixels trailing_margin, CSSPixels content_width, CSSPixels content_height, CSSPixels border_box_top, CSSPixels border_box_bottom, RefPtr<Gfx::GlyphRun> glyph_run = {});
+    void add_fragment(Node const& layout_node, int start, int length, CSSPixels leading_size, CSSPixels trailing_size, CSSPixels leading_margin, CSSPixels trailing_margin, CSSPixels content_width, CSSPixels content_height, CSSPixels border_box_top, CSSPixels border_box_bottom, RefPtr<Gfx::GlyphRun> glyph_run = {}, CSS::Direction direction = CSS::Direction::Ltr);
```
Comment (diegoiast) [inline Userland/Libraries/LibWeb/Layout/LineBox.h:24]: This default is wrong. You should use the default direction of the document. For example in HTML4:

```
<html dir=rtl">
<body>
aa....!!!
```

The "!!!" should be at the left of the "aa".

File: Userland/Libraries/LibWeb/Layout/LineBoxFragment.cpp [hunk 3]
```diff
@@ -45,4 +45,35 @@ bool LineBoxFragment::is_atomic_inline() const
...
+        break;
+    case CSS::Direction::Rtl:
+        for (auto& glyph : m_glyph_run->glyphs()) {
+            glyph.visit([&](auto& glyph) {
+                if (glyph.position.x() >= m_rtl_insert_position)
+                    glyph.position.translate_by(run_width.to_float(), 0);
+            });
+        }
+        for (auto& glyph : glyph_run->glyphs()) {
+            glyph.visit([&](auto& glyph) { glyph.position.translate_by(m_rtl_insert_position, 0); });
```
Comment (diegoiast) [inline Userland/Libraries/LibWeb/Layout/LineBoxFragment.cpp:71]: Please test this with mixed paragraphs, which contain LTR and RTL characters. In the text bellow the aleph (looks like an X) is first, and the mem (more or less like an O) should be on the end. In the middle - you should have perfect english. I will leave the honor to you to make proper test cases :)

Both of the next examples contain the same text, but different direction of the paragraph (emulated using quote and code in markdown).

```
אאא test מממ
```

> אאא test מממ

**Review by diegoiast — COMMENTED** [review]
There are some issues with the implementation. Might be my limited understanding of the API.

File: Userland/Libraries/LibWeb/Layout/InlineFormattingContext.cpp (hunk 1 above)
Comment (BenJilks) [reply Userland/Libraries/LibWeb/Layout/InlineFormattingContext.cpp]: Yeah, this logic was a bit off, we need to know both directions. The layout direction is now passed into the line builder on construction, and each glyph run contains its own direction value.

File: Userland/Libraries/LibWeb/Layout/LineBox.h (hunk 2 above)
Comment (BenJilks) [reply Userland/Libraries/LibWeb/Layout/LineBox.h]: We now inherit the default direction, from the context box's computed CSS value.

File: Userland/Libraries/LibWeb/Layout/LineBoxFragment.cpp (hunk 3 above)
Comment (BenJilks) [reply Userland/Libraries/LibWeb/Layout/LineBoxFragment.cpp]: I've created a test case covering different combinations of mixing LRT with RTL, and have got all of them to behave the same as Firefox and Chrome:

```html
<div dir="ltr" align="left">hello test 1, 2, 3!</div>
<div dir="rtl" align="left">hello test 1, 2, 3!</div>
<div dir="ltr" align="right">hello test 1, 2, 3!</div>
<div dir="rtl" align="right">hello test 1, 2, 3!</div>

<div dir=ltr>אא aaa bbb ccc מממ</div>
<div dir=rtl>אא aaa bbb ccc מממ</div>

<div dir=ltr>אא 1 2 3 מממ</div>
<div dir=rtl>אא 1 2 3 מממ</div>

<div dir=ltr>aa....!!!</div>
<div dir=rtl>aa....!!!</div>

<div dir=ltr>حسنًا ، hello friends مرحباً أيها ا test لأصدقاء end</div>
<div dir=rtl>حسنًا ، hello friends مرحباً أيها ا test لأصدقاء end</div>
```

There is one case I know about that doesn't work, and that's brackets. It should replace them with their mirrored counterparts when flowing the text RTL. But LibUnicode doesn't currently implement mirrored characters, so I feel that might be outside of the PR.

If you know of anything big I missed, please let me know. Just note that for this PR, I'm purely focusing on the arrangement withing a single text fragment. Anything outside of that will effect layout and I feels that's beyond the scope.

**Comment by diegoiast** [comment]
Tested on 263dcde888 :

Found another use case which fails: you need to re-order the chunks - also between html element. For example, if a sentence has a `<a href>` this breaks the order. See this example (ignore the missing fonts... the problem is still illustrated), Firefox correct - ton the top, Ladybird wrong - on the bottom:

image

I used this example:
```html
<div dir=ltr align=left>
<span style="background: #ffaabb">אאאא</span>
<span style="background: #aabbff">1111</span>
<span style="background: #ff22bb">aaaa</span>
</div>

<div dir=ltr align=right>
<span style="background: #ffaabb">אאאא</span>
<span style="background: #aaffbb">1111</span>
<span style="background: #ff22bb">aaaa</span>
</div>


<div dir=rtl align=right>
<span style="background: #ffaabb">אאאא</span>
<span style="background: #aaffbb">1111</span>
<span style="background: #ff22bb">aaaa</span>
</div>


<div dir=rtl align=left>
<span style="background: #ffaabb">אאאא</span>
<span style="background: #aaffbb">1111</span>
<span style="background: #ff22bb">aaaa</span>
</div>
```

Some other issues still available:

- Hypens are shown at the wrong place:  
image
- Brackets should be reversed (sometimes? how does one define a left bracket, and an open bracket? those are different in RTL contents)

**Comment by awesomekling** [comment]
Seems like a solid progression! Of course there is more work to do, but let's not block this PR forever :)

**Pull request merged**